docs/Driver.md

```diff
@@ -0,0 +1,311 @@
+temporary directory. Trying to specify the outputs for every input file on the
+command line would be awkward and ridiculous, so instead we use an _output file
+map._
+
+An output file map contains a JSON dictionary that looks like this:
+
+```json
+{
+  "/path/to/src/foo.swift": {
+    "object": "/path/to/build/foo.o",
+    "swift-dependencies": "/path/to/build/foo.swiftdeps",
+    "diagnostics": "/path/to/build/foo.dia",
+  },
+  "/path/to/src/bar.swift": {
+    "object": "/path/to/build/bar.o"
+    "swift-dependencies": "/path/to/build/bar.swiftdeps",
+    "diagnostics": "/path/to/build/bar.dia",
+  },
+  "": {
+    "swift-dependencies": "/path/to/build/main-build-record.swiftdeps",
+  }
+}
+```
+
+The build system is responsible for generating this file. The input file names
+don't have to be absolute paths, but they do have to match the form used in the
+invocation of the compiler. No canonicalization is performed. The
+"swift-dependencies" entries are required to perform incremental builds (see
+below). The "diagnostics" entries are only for your own use; if you don't need
+them they can be omitted.
+
+The output file map accepts other entries, but they should not be considered
+stable. Please stick to what's shown here.
+
+(It *would* be nice to have a way to get reasonable default behavior by
+pointing `-o` at a directory, or by adding an `-intermediate-directory`
+argument or something. Patches welcome.)
+
+
+## Whole-Module Optimization ##
+
+When the `-whole-module-optimization` flag is passed to Swift, the compilation
+model changes significantly. In this mode, the driver only invokes the frontend
+once, and there is no primary file. Instead, every file is parsed and
+type-checked, and then all generated code is optimized together.
+
+Whole-module builds actually have two modes: threaded and non-threaded. The
+default is non-threaded, which produces a single object file for the entire
+compilation unit (.o). If you're just producing object files, you can use the
+usual `-o` option with `-c` to control where the single output goes.
+
+In threaded mode, one object file is produced for every input source file, and
+the "backend" processing of the generated code (basically, everything that's
+not Swift-language-specific) is performed on multiple threads. Like
+non-whole-module compilation, the locations of the object files is controlled
+by the output file map. Only one file is produced for each non-object output,
+however. The location of these additional outputs is controlled by command-line
+options, except for serialized diagnostics, which use an entry under `""` in
+the output file map.
+
+Threaded mode is controlled by the `-num-threads` command-line option rather
+than `-j` used to control the number of jobs (simultaneous subprocesses spawned
+by the driver). Why? Well, they're not quite setting the same thing...but it's
+unclear whether that difference is buying us anything.
```

Comment:
Could you elaborate? As a reader, I would have expected the next paragraph to go into detail on why you thought the difference didn't have any benefits. As is, on the other hand, I just have unanswered questions lingering in my brain.

Perhaps pare this down to just:

> Threaded mode is controlled by the `-num-threads` command-line option rather than `-j`, which is used to control the number of jobs (simultaneous subprocesses spawned by the driver).